### DISCUSSION TITLE
Alpine does not see the `x-show` attribute on elements swapped in by htmx

### DISCUSSION THREAD

**Discussion by adhusson** [alpinejs/alpine#3809] [1. Help]
I'm using htmx & Alpine together and I'm having an issue with htmx-swapped content. 

I have an inline edit form that looks like this:

```
<div hx-target="this" hx-swap="outerHTML" x-data="{editing: false, content:'My comment'}">
  <button x-show="!editing" x-on:click="editing=true">🖊</button>
  <button x-show="editing" x-on:click="editing=false">↩️️</button>

  <div x-show="!editing" x-text="content"></div>
  <form x-cloak x-show="editing" id="342" hx-put="/comment/342">
    <input x-bind:value="content" name="content" type="text" />
    <input type="submit" value="go">
  </form>
</div>
```

When the server receives `PUT /comment/342`, it re-renders the snippet above but with an updated content value.

My problem is that the swapped content shows both the content itself  (`<div x-show="!editing" x-text="content"></div>`) and the edit form (`<form>...</form>`). 

It's as if Alpine is completely ignoring the `x-show` attributes of the swapped content. But it does process the inserted elements in some way (for instance it removes the `x-cloak` attribute from the form)

**Comment by SimoTod** [comment]
I don't use htmx but from the syntax it seems like you are swapping the outerHTML (so you are also removing the element with x-data and it's data scope where you defined editing)

**Reply by adhusson** [reply]
I am swapping the outerHTML, but the newly swapped-in `<div>` also has the `x-data` and other attributes.

**Reply by SimoTod** [reply]
Yeah but Alpine creates a real js object containing the scope and attaches it to a property of that dom element (html is just a textual version of the actual dom tree). If you change the element for a different one, the new one won't have tge scope attached anymore and because it looks exactly the same, Alpine may struggle to understand that it should recompute x-data again. This is one of the reasons why the morph plugin was built.
You can try adding x-data to an external div and just swap the content.
Otherwise htmx has a alpine morph plugin which should probably fix your problem: https://htmx.org/docs/#swapping

**Reply by adhusson** [reply]
Thank you. `innerHTML`swap works but requires additional server-side logic to separate rendering the comment with + without the container. You also have to listen to the htmx response and set `editing` to `false`.

alpine-morph partially works because it interprets `x-show`, but it does not interpret the `x-data` attribute of the swapped-in element. You still need to listen to the htmx response and set `editing` to `false`.

**Reply by SimoTod** [reply]
Yeah, I think the latter is expected. The morph will diff the html and keep the current state (the html attribute is just a textual representation of the state and doesn't override the state once Alpine initialise it) so editing needs to be set to false by you.

**Reply by ekwoka** [reply]
You shouldn't need to change serverside


You can tell htmx the target element to use in the returned html.

Additionally, you can use hx-swap="morph" with the morph plugin to have Alpine handle making updates to the Dom and preserving state.

**Reply by adhusson** [reply]
Thanks, yes `hx-swap="morph"` almost fully works, but alpine does not automatically interpret the `x-data` attribute of the swapped-in element (it does automatically interpret the `x-show` attribute of the swapped-in element's children).

**Reply by ekwoka** [reply]
It's not meant to. Would the swapped in element have a different x-data? x-data won't be reevaluated, as generally that would reset the component, which is what morph is meant NOT to do.

**Reply by adhusson** [reply]
The idea was to swap in a copy of the element with `editing` set to false, thus hiding the input field and showing the div with the updated comment. 

I think it would be a nice pattern: by sending back the updated comment fragment, you reset it to its exact state before the user started to edit it.

Another way to look at it is that the application state is both the DOM state and the javascript state, so if a swapped-in HTML fragment has data that is semantically meaningful for the javascript state, that should be taken into account and merged in.

**Reply by SimoTod** [reply]
That's make sense in your use case but it would break a lot of existing stuff (anything using Livewire probably). The majority of the times people send back the original template but doesn't expect Alpine to reset the state. At least, that's what morphing does (you can also see that your x-data attribute doesn't normally change once alpine is initialised, the state does). 

In theory, you should probably set editing back to false when you send the form to avoid race conditions between concurrent editing.

**Reply by ekwoka** [reply]
On that note, a new x-init attribute would be evaluated.

That could be used to change the state in progress.

Also, just htmx events

**Comment by danielnieto** [comment]
@adhusson I had the same issue as you, it seems it's more of a problem on HTMX side, after a lot of time I remembered that HTMX tries to keep the same DOM element if it has a "stable" ID between swaps, in my case I had the following:

`<div role="menu" x-cloak x-show="isOpen" id="contact-menu-{{ contact.id }}" @click.outside="isOpen=false" @keydown.escape.window="isOpen=false">
`

and I was seeing the same issue you described, first time it was fine but as soon as the page got re-rendered (by HTMX) all of my menus would be visible! I fixed it by removing the `id` from the element and that solved the problem... Looks like HTMX trying to preserve the DOM element and just mutate it messes up the interaction with Alpine

**Reply by danielnieto** [reply]
or keep the id and add `hx-preserve`, it also works:

`<div role="menu" hx-preserve x-cloak x-show="isOpen" id="contact-menu-{{ contact.id }}" @click.outside="isOpen=false" @keydown.escape.window="isOpen=false">`

**Reply by ekwoka** [reply]
It is helpful to not use `x-cloak` but instead use `style="display:none;"` on things expected for x-show to hide on load. This is a way to solve these kinds of issues.

**Reply by steven-mercatante** [reply]
+1 for manually adding `style="display:none"`. This fixed the HTMX + Alpine issue for me.

**Reply by jh0ker** [reply]
This issue specifically is related to the settling mechanism in HTMX. See [this issue in the HTMX repo](https://github.com/bigskysoftware/htmx/issues/412) for more info. It also explains why these workarounds work and discusses some other ones.
